src/libstd/sys/unix/condvar.rs

```diff
@@ -8,22 +8,51 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+use prelude::v1::*;
+
 use cell::UnsafeCell;
 use libc;
-use ptr;
 use sys::mutex::{self, Mutex};
-use time::{Instant, Duration};
+use time::Duration;
 
 pub struct Condvar { inner: UnsafeCell<libc::pthread_cond_t> }
 
 unsafe impl Send for Condvar {}
 unsafe impl Sync for Condvar {}
 
+
 impl Condvar {
-    pub const fn new() -> Condvar {
-        // Might be moved and address is changing it is better to avoid
-        // initialization of potentially opaque OS data before it landed
-        Condvar { inner: UnsafeCell::new(libc::PTHREAD_COND_INITIALIZER) }
+
+    #[cfg(any(target_os = "macos", target_os = "ios"))]
+    pub fn new() -> Box<Condvar> {
+        box Condvar { inner: UnsafeCell::new(libc::PTHREAD_COND_INITIALIZER) }
+    }
+
+    #[cfg(not(any(target_os = "macos", target_os = "ios")))]
+    pub fn new() -> Box<Condvar> {
+        use mem;
+
+        // Holds a mutable reference to initialized pthread_condattr_t
+        // and ensures that pthread_condattr_destroy is called when
+        // leaving the scope.
+        struct CondattrHolder<'a>(&'a mut libc::pthread_condattr_t);
+
+        impl<'a> Drop for CondattrHolder<'a> {
+            fn drop(&mut self) {
+                unsafe { libc::pthread_condattr_destroy(self.0) };
+            }
+        }
```

Comment:
I think it's fine to not worry about assertions below cleaning this up, this can just be inserted after the `pthread_cond_init` (like mutex initialization)